### PR TITLE
Introduce basic helm chart

### PR DESCRIPTION
**Pull request by wilfriedroset**
# What does this PR do?

This PR introduce a basic helm chart to deploy TGI on k8s. Deploying TGI is not impossible without the helm chart but I reckon that having one will facilitate the overall adoption.

The chart is composed by 3 resources: service, statefulset and an optional  ingress.

My starting point was the following  k8s manifest:
```yaml
---
apiVersion: apps/v1
kind: StatefulSet
metadata:
  name: text-generation-inference
  namespace: default
  labels:
    app: text-generation-inference
spec:
  replicas: 1
  selector:
    matchLabels:
      app: text-generation-inference
  template:
    metadata:
      name: text-generation-inference
      labels:
        app: text-generation-inference
    spec:
      containers:
        - name: text-generation-inference
          image: ghcr.io/huggingface/text-generation-inference:0.9.3
          imagePullPolicy: IfNotPresent
          ports:
            - containerPort: 80
          command:
            - "text-generation-launcher"
            - "--model-id"
            - "bigcode/starcoderbase-7b"
            - "--num-shard"
            - "1"
            - "--max-batch-total-tokens"
            - "16000"
            - "--max-total-tokens"
            - "2048"
            - "--trust-remote-code"
            - "--max-concurrent-requests"
            - "10"
            - "--revision"
            - "4ab631381edb607557cbb04b6e9a225bad16807c"
          env:
            - name: USE_FLASH_ATTENTION
              value: "false"
            - name: HUGGING_FACE_HUB_TOKEN
              value: "hf_FIXME"
            - name: HF_HUB_ENABLE_HF_TRANSFER
              value: "false"
          resources:
            limits:
              nvidia.com/gpu: "1"
            requests:
              nvidia.com/gpu: "1"
          volumeMounts:
            - mountPath: "/data"
              name: starcoder
      volumes:
        - name: starcoder
          persistentVolumeClaim:
            claimName: starcoder
      restartPolicy: Always
  volumeClaimTemplates:
    - metadata:
        name: starcoder
      spec:
        accessModes:
          - ReadWriteOnce
        resources:
          requests:
            storage: 150Gi
        storageClassName: csi-cinder-high-speed
---
apiVersion: v1
kind: Service
metadata:
  name: text-generation-inference
  namespace: default
  labels:
    app: text-generation-inference
spec:
  selector:
    app: text-generation-inference
  ports:
    - protocol: TCP
      port: 80
      targetPort: 80
      name: http
  type: ClusterIP
---
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: text-generation-inference
  namespace: default
  annotations:
    nginx.ingress.kubernetes.io/rewrite-target: /
    cert-manager.io/cluster-issuer: "selfsigned"
spec:
  ingressClassName: nginx
  rules:
    - host: starcoder.[redacted]
      http:
        paths:
          - path: /
            pathType: Prefix
            backend:
              service:
                name: text-generation-inference
                port:
                  number: 80
  tls:
    - hosts:
        - starcoder.[redacted]
      secretName: starcoder-tls-secret
```

I've manually tested the chart with the following  values:
```yaml
---
args:
  - "--model-id"
  - "bigcode/starcoderbase-7b"
  - "--num-shard"
  - "1"

env:
  HUGGING_FACE_HUB_TOKEN: hf_FIXME

persistence:
  storageClassName: "csi-cinder-high-speed"
  storage: 150Gi

ingress:
  enabled: true
  className: "nginx"
  hosts:
    - host: startcoder.[redacted]
      paths:
        - path: /
          pathType: "Prefix"
  tls:
    - secretName: starcoder-tls
      hosts:
        - startcoder.[redacted]
```

Without the ingress:
```
❯ helm install --dry-run -f /tmp/starcoder.yaml startcoder .
NAME: startcoder
LAST DEPLOYED: Thu Nov  9 21:25:46 2023
NAMESPACE: default
STATUS: pending-install
REVISION: 1
TEST SUITE: None
HOOKS:
MANIFEST:
---
# Source: text-generation-inference/templates/service.yaml
apiVersion: v1
kind: Service
metadata:
  name: startcoder-text-generation-inference
  namespace: default
  labels:
    helm.sh/chart: text-generation-inference-0.1.0
    app.kubernetes.io/name: text-generation-inference
    app.kubernetes.io/instance: startcoder
    app.kubernetes.io/version: "1.1.0"
    app.kubernetes.io/managed-by: Helm
spec:
  type: ClusterIP
  ports:
    - port: 80
      targetPort:
      protocol: TCP
      name: http
  selector:
    app.kubernetes.io/name: text-generation-inference
    app.kubernetes.io/instance: startcoder
---
# Source: text-generation-inference/templates/statefulset.yaml
apiVersion: apps/v1
kind: StatefulSet
metadata:
  name: startcoder-text-generation-inference
  namespace: default
  labels:
    helm.sh/chart: text-generation-inference-0.1.0
    app.kubernetes.io/name: text-generation-inference
    app.kubernetes.io/instance: startcoder
    app.kubernetes.io/version: "1.1.0"
    app.kubernetes.io/managed-by: Helm
spec:
  replicas: 1
  serviceName: text-generation-inference
  selector:
    matchLabels:
      app.kubernetes.io/name: text-generation-inference
      app.kubernetes.io/instance: startcoder
  template:
    metadata:
      name: text-generation-inference
      labels:
        app.kubernetes.io/name: text-generation-inference
        app.kubernetes.io/instance: startcoder
    spec:
      containers:
        - name: text-generation-inference
          image: ghcr.io/huggingface/text-generation-inference:1.1.0
          imagePullPolicy:
          ports:
            - containerPort: 80
          command:
            - "text-generation-launcher"
            - "--model-id"
            - "bigcode/starcoderbase-7b"
            - "--num-shard"
            - "1"
          env:
            - name: HUGGING_FACE_HUB_TOKEN
              value: "hf_FIXME"
          volumeMounts:
            - mountPath: "/data"
              name: storage
      volumes:
        - name: storage
          persistentVolumeClaim:
            claimName: storage
      restartPolicy: Always
  volumeClaimTemplates:
    - metadata:
        name: storage
      spec:
        accessModes:
          - "ReadWriteOnce"
        storageClassName: default
        resources:
          requests:
            storage: 150Gi

NOTES:
1. Get the application URL by running these commands:
  export POD_NAME=$(kubectl get pods --namespace default -l "app.kubernetes.io/name=text-generation-inference,app.kubernetes.io/instance=startcoder" -o jsonpath="{.items[0].metadata.name}")
  export CONTAINER_PORT=$(kubectl get pod --namespace default $POD_NAME -o jsonpath="{.spec.containers[0].ports[0].containerPort}")
  echo "Visit http://127.0.0.1:8080 to use your application"
  kubectl --namespace default port-forward $POD_NAME 8080:$CONTAINER_PORT
```

With the ingress enabled:
```
❯ helm install --dry-run -f /tmp/starcoder.yaml startcoder .
NAME: startcoder
LAST DEPLOYED: Thu Nov  9 21:28:37 2023
NAMESPACE: default
STATUS: pending-install
REVISION: 1
TEST SUITE: None
HOOKS:
MANIFEST:
---
# Source: text-generation-inference/templates/service.yaml
apiVersion: v1
kind: Service
metadata:
  name: startcoder-text-generation-inference
  namespace: default
  labels:
    helm.sh/chart: text-generation-inference-0.1.0
    app.kubernetes.io/name: text-generation-inference
    app.kubernetes.io/instance: startcoder
    app.kubernetes.io/version: "1.1.0"
    app.kubernetes.io/managed-by: Helm
spec:
  type: ClusterIP
  ports:
    - port: 80
      targetPort:
      protocol: TCP
      name: http
  selector:
    app.kubernetes.io/name: text-generation-inference
    app.kubernetes.io/instance: startcoder
---
# Source: text-generation-inference/templates/statefulset.yaml
apiVersion: apps/v1
kind: StatefulSet
metadata:
  name: startcoder-text-generation-inference
  namespace: default
  labels:
    helm.sh/chart: text-generation-inference-0.1.0
    app.kubernetes.io/name: text-generation-inference
    app.kubernetes.io/instance: startcoder
    app.kubernetes.io/version: "1.1.0"
    app.kubernetes.io/managed-by: Helm
spec:
  replicas: 1
  serviceName: text-generation-inference
  selector:
    matchLabels:
      app.kubernetes.io/name: text-generation-inference
      app.kubernetes.io/instance: startcoder
  template:
    metadata:
      name: text-generation-inference
      labels:
        app.kubernetes.io/name: text-generation-inference
        app.kubernetes.io/instance: startcoder
    spec:
      containers:
        - name: text-generation-inference
          image: ghcr.io/huggingface/text-generation-inference:1.1.0
          imagePullPolicy:
          ports:
            - containerPort: 80
          command:
            - "text-generation-launcher"
            - "--model-id"
            - "bigcode/starcoderbase-7b"
            - "--num-shard"
            - "1"
          env:
            - name: HUGGING_FACE_HUB_TOKEN
              value: "hf_FIXME"
          volumeMounts:
            - mountPath: "/data"
              name: storage
      volumes:
        - name: storage
          persistentVolumeClaim:
            claimName: storage
      restartPolicy: Always
  volumeClaimTemplates:
    - metadata:
        name: storage
      spec:
        accessModes:
          - "ReadWriteOnce"
        storageClassName: default
        resources:
          requests:
            storage: 150Gi
---
# Source: text-generation-inference/templates/ingress.yaml
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: startcoder-text-generation-inference
  labels:
    helm.sh/chart: text-generation-inference-0.1.0
    app.kubernetes.io/name: text-generation-inference
    app.kubernetes.io/instance: startcoder
    app.kubernetes.io/version: "1.1.0"
    app.kubernetes.io/managed-by: Helm
spec:
  ingressClassName: nginx
  tls:
    - hosts:
        - "startcoder.[redacted]"
      secretName: starcoder-tls
  rules:
    - host: "startcoder.[redacted]"
      http:
        paths:
          - path: /
            pathType: Prefix
            backend:
              service:
                name: startcoder-text-generation-inference
                port:
                  number: 80

NOTES:
1. Get the application URL by running these commands:
  https://startcoder.[redacted]/
```

```
❯ helm install -f /tmp/starcoder.yaml startcoder .
NAME: startcoder
LAST DEPLOYED: Thu Nov  9 21:30:11 2023
NAMESPACE: default
STATUS: deployed
REVISION: 1
TEST SUITE: None
NOTES:
1. Get the application URL by running these commands:
  https://startcoder.[redacted]/
```

Then we can check the different resources:
```
❯ k get services -o wide
NAME                                   TYPE        CLUSTER-IP    EXTERNAL-IP   PORT(S)                        AGE    SELECTOR
kubelet                                ClusterIP   None          <none>        10250/TCP,10255/TCP,4194/TCP   2d4h   <none>
kubernetes                             ClusterIP   10.3.0.1      <none>        443/TCP                        2d4h   <none>
startcoder-text-generation-inference   ClusterIP   10.3.67.238   <none>        80/TCP                         33s    app.kubernetes.io/instance=startcoder,app.kubernetes.io/name=text-generation-inference
```
```
❯ k get ingress -o wide
NAME                                   CLASS   HOSTS                                   ADDRESS         PORTS     AGE
startcoder-text-generation-inference   nginx   startcoder.[redacted]   [redacted]   80, 443   54s
```
```
❯ k get statefulset -o wide
NAME                                   READY   AGE   CONTAINERS                  IMAGES
startcoder-text-generation-inference   0/1     69s   text-generation-inference   ghcr.io/huggingface/text-generation-inference:1.1.0
```
```
❯ k get pod -o wide
NAME                                     READY   STATUS    RESTARTS   AGE   IP           NODE                              NOMINATED NODE   READINESS GATES
startcoder-text-generation-inference-0   1/1     Running   0          46s   10.2.10.20   data-plane-t2-le-45-node-84eeea   <none>           <none>
```
```
❯ k get pvc -o wide
NAME                                             STATUS   VOLUME                                                                   CAPACITY   ACCESS MODES   STORAGECLASS            AGE   VOLUMEMODE
storage-startcoder-text-generation-inference-0   Bound    ovh-managed-kubernetes-ugeiv2-pvc-0cfad25a-ef16-4041-9cbc-e78e40d38b46   150Gi      RWO            csi-cinder-high-speed   64s   Filesystem
```

## Who can review?

@OlivierDehaene OR @Narsil OR @Hugoch OR @McPatate

### CODE REVIEW

File: helm/charts/text-generation-inference/Chart.yaml [hunk 1]
```diff
@@ -0,0 +1,29 @@
+apiVersion: v2
```
Comment (wilfriedroset) [inline helm/charts/text-generation-inference/Chart.yaml:1]: helm lint advise to add an icon, if you have one for TGI I can amend the PR do add it.

File: helm/charts/text-generation-inference/Chart.yaml [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+home: https://github.com/huggingface/text-generation-inference
+annotations:
+  "artifacthub.io/license": HFOILv1.0
+  "artifacthub.io/links": |
+    - name: Upstream Project
+      url: https://github.com/huggingface/text-generation-inference
+    - name: Documentation
+      url: https://huggingface.co/docs/text-generation-inference/index
+maintainers:
+  - name: wilfriedroset
```
Comment (wilfriedroset) [inline helm/charts/text-generation-inference/Chart.yaml:17]: Feel free to remove my name here, no worries.

File: helm/charts/text-generation-inference/Chart.yaml [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+keywords:
```
Comment (wilfriedroset) [inline helm/charts/text-generation-inference/Chart.yaml:18]: those keywords are the same as the one associated to TGI github repository

File: helm/charts/text-generation-inference/values.yaml [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+resources:
+  limits:
+    nvidia.com/gpu: "1"
+  requests:
+    nvidia.com/gpu: "1"
+
+args:
+  []
+  # See: https://huggingface.co/docs/text-generation-inference/basic_tutorials/launcher
+  # - "--model-id"
```
Comment (wilfriedroset) [inline helm/charts/text-generation-inference/values.yaml:25]: we could alternatively provide a simpler way to define the model to use, example:
```
model:
  id: bigcode/starcoderbase-7b
  revision: 4ab631381edb607557cbb04b6e9a225bad16807c
```
This way  user do not need to know the exact args to provide.

File: helm/charts/text-generation-inference/values.yaml [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+  # - "bigcode/starcoderbase-7b"
+  # - "--revision"
+  # - "4ab631381edb607557cbb04b6e9a225bad16807c"
+  # - "--num-shard"
+  # - "1"
```
Comment (wilfriedroset) [inline helm/charts/text-generation-inference/values.yaml:30]: Similarly to model-id, we could have an higher level value `num-shard: 1` and pass it directly to the args list.
This  value  could be use for the resources as well.

File: helm/charts/text-generation-inference/values.yaml [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+          #   service:
+          #     name: text-generation-inference
+          #     port:
+          #       number: 80
+  tls: []
+  #  - secretName: chart-example-tls
+  #    hosts:
+  #      - chart-example.local
+
+persistence:
```
Comment (wilfriedroset) [inline helm/charts/text-generation-inference/values.yaml:59]: What do you think about those default values?
Should we go for a bigger storage? `ReadWriteMany` by default?

File: helm/charts/text-generation-inference/templates/ingress.yaml [hunk 7]
```diff
@@ -0,0 +1,61 @@
+{{- if .Values.ingress.enabled -}}
+{{- $fullName := include "text-generation-inference.fullname" . -}}
+{{- $svcPort := .Values.service.port -}}
+{{- if and .Values.ingress.className (not (semverCompare ">=1.18-0" .Capabilities.KubeVersion.GitVersion)) }}
+  {{- if not (hasKey .Values.ingress.annotations "kubernetes.io/ingress.class") }}
+  {{- $_ := set .Values.ingress.annotations "kubernetes.io/ingress.class" .Values.ingress.className}}
+  {{- end }}
+{{- end }}
+{{- if semverCompare ">=1.19-0" .Capabilities.KubeVersion.GitVersion -}}
```
Comment (McPatate) [inline helm/charts/text-generation-inference/templates/ingress.yaml:9]: Haven't you specified in the Chart.yaml that it is compatible for k8s ^1.27 ?
(not saying we should change the code here, rather update the kubeversion in the Chart.yaml)

File: helm/charts/text-generation-inference/templates/statefulset.yaml [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+            - containerPort: 80
+          command:
+            - "text-generation-launcher"
+            {{- range .Values.args }}
+            - {{ . | quote }}
+            {{- end }}
+          env:
+          {{- range $key, $value := .Values.env }}
+            - name: {{ $key }}
+              value: {{ $value|quote }}
```
Comment (McPatate) [inline helm/charts/text-generation-inference/templates/statefulset.yaml:34]:
```suggestion
              value: {{ $value | quote }}
```

File: helm/charts/text-generation-inference/values.yaml (hunk 4 above)
Comment (McPatate) [reply helm/charts/text-generation-inference/values.yaml:25]: Intuitively I would have done it like you're proposing, but args doesn't change much and is closer to the real thing, no strong opinion on my side.

File: helm/charts/text-generation-inference/templates/service.yaml [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+kind: Service
+metadata:
+  name: {{ include "text-generation-inference.fullname" . }}
+  namespace: {{ .Release.Namespace }}
+  labels:
+    {{- include "text-generation-inference.labels" . | nindent 4 }}
+spec:
+  type: {{ .Values.service.type }}
+  ports:
+    - port: {{ .Values.service.port }}
```
Comment (McPatate) [inline helm/charts/text-generation-inference/templates/service.yaml:11]: Is the port for metrics the same as the port for requests ?

**Review by McPatate — COMMENTED** [review]
Great job @wilfriedroset !

I'm seeing two things you could add:
- an HPA (maybe that's a bit complicated since it'd require to configure a prom adapter and scaling on a custom metric rather than CPU/GPU util)
- a PDB

File: helm/charts/text-generation-inference/templates/ingress.yaml (hunk 7 above)
Comment (wilfriedroset) [reply helm/charts/text-generation-inference/templates/ingress.yaml:9]: the ingress definition is straight from `helm create`. the default definition is enough for the current state of the chart.
I've define k8s ^1.27 in Chart.yaml as this is the version I'm currently using for testing.
I can align everything to 1.27 if that makes more sense

File: helm/charts/text-generation-inference/templates/service.yaml (hunk 9 above)
Comment (wilfriedroset) [reply helm/charts/text-generation-inference/templates/service.yaml:11]: I believe they are, you need to call /generate at least once to get the metrics populated.
```
❯ curl -s -H "Content-type: application/json"  -k  https://startcoder.[redacted]/generate -d @/tmp/foo.json | jq

{
  "generated_text": "\nimport sys\nimport time\n\nfrom. import config\nfrom. import utils\nfrom."
}
❯ curl -sk https://startcoder.[redacted]/metrics | head
# TYPE tgi_request_success counter
tgi_request_success 3

# TYPE tgi_batch_inference_success counter
tgi_batch_inference_success{method="prefill"} 3
tgi_batch_inference_success{method="decode"} 57

# TYPE tgi_request_count counter
tgi_request_count 3
```

**Comment by wilfriedroset** [comment]
> an HPA (maybe that's a bit complicated since it'd require to configure a prom adapter and scaling on a custom metric rather than CPU/GPU util)

I've puporsedly remove the default autoscaling block. As you pointed out, building an HPA for tgi requires additional setup. I want test first have a basic chart merge and use it as a building block for the next step.

> a PDB

I've amended with a PDB.

**Comment by wilfriedroset** [comment]
This is not stale, I'm waiting for maintainers' feedbacks.

**Comment by McPatate** [comment]
cc @OlivierDehaene @Narsil

**Pull request closed without merging**